File: 0100-same-tree/0100-same-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean result = true;
    
    public boolean isSameTree(TreeNode p, TreeNode q) {
        // approach: do DFS on both trees at the same time
        
        dfs(p, q);
        return result;
    }
    
    public void dfs(TreeNode p, TreeNode q) {
        if (p == null || q == null) {
            if (!(p == null && q == null)) {
                result = false;
            }
            
            return;
        }
        
        dfs(p.left, q.left);
        dfs(p.right, q.right);
        
        if (p.val != q.val) {
            result = false;
        }
    }
}

File: 0100-same-tree/0100-same-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # approach: top-down DFS, compare node values
        
        result = True
        
        def helper(p, q):
            nonlocal result
            
            if not p and not q:
                return
            
            if not p or not q:
                result = False
                return
            
            if p.val != q.val:
                result = False
            
            helper(p.left, q.left)
            helper(p.right, q.right)
        
        helper(p, q)
        return result
            